src/core/dyad_core.c

```diff
@@ -0,0 +1,622 @@
+            kvs_namespace,
+            strlen(kvs_namespace)+1
+           );
+    printf("Setting prod_managed_path\n");
+    if (prod_managed_path == NULL)
+    {
+        (*ctx)->prod_managed_path = NULL;
+    }
+    else
+    {
+        (*ctx)->prod_managed_path = (char*) malloc(strlen(prod_managed_path)+1);
+        if ((*ctx)->prod_managed_path == NULL)
+        {
+            fprintf (stderr, "Could not allocate buffer for Producer managed path!\n");
+            free((*ctx)->kvs_namespace);
+            free(*ctx);
+            *ctx = NULL;
+            return DYAD_NOCTX;
+        }
+        strncpy(
+                (*ctx)->prod_managed_path,
+                prod_managed_path,
+                strlen(prod_managed_path)+1
+            );
+    }
+    printf("Setting cons_managed_path\n");
+    if (cons_managed_path == NULL)
+    {
+        (*ctx)->cons_managed_path = NULL;
+    }
+    else
+    {
+        (*ctx)->cons_managed_path = (char*) malloc(strlen(cons_managed_path)+1);
+        if ((*ctx)->cons_managed_path == NULL)
+        {
+            fprintf (stderr, "Could not allocate buffer for Consumer managed path!\n");
+            free((*ctx)->kvs_namespace);
+            free((*ctx)->prod_managed_path);
+            free(*ctx);
+            *ctx = NULL;
+            return DYAD_NOCTX;
+        }
+        strncpy(
+                (*ctx)->cons_managed_path,
+                cons_managed_path,
+                strlen(cons_managed_path)+1
+            );
+    }
+    if ((*ctx)->prod_managed_path == NULL && (*ctx)->cons_managed_path == NULL)
+    {
+        fprintf (stderr, "Warning: DYAD instance is neither producer nor consumer!\n");
+    }
+    printf("Opening Flux\n");
+    (*ctx)->reenter = true;
+    (*ctx)->h = flux_open(NULL, 0);
+    if ((*ctx)->h == NULL)
+    {
+        fprintf (stderr, "Could not open Flux handle!\n");
+        return DYAD_FLUXFAIL;
+    }
+    printf("Getting Flux Rank\n");
+    if (flux_get_rank((*ctx)->h, &((*ctx)->rank)) < 0)
+    {
+        fprintf(stderr, "Could not get Flux rank\n");
+        FLUX_LOG_ERR ((*ctx)->h, "Could not get Flux rank!\n");
+        return DYAD_FLUXFAIL;
+    }
+    printf("Initialization complete\n");
+    (*ctx)->initialized = true;
+    // TODO Print logging info
+    return DYAD_OK;
+}
+
+int dyad_produce(dyad_ctx_t *ctx, const char *fname)
+{
+    int rc = validate_context(ctx, false, true);
+    if (DYAD_IS_ERROR(rc))
+    {
+        return rc;
+    }
+    return dyad_commit(ctx, fname);
+}
+
+int dyad_consume(dyad_ctx_t *ctx, const char *fname)
+{
+    int rc = validate_context(ctx, true, false);
+    if (DYAD_IS_ERROR(rc))
+    {
+        return rc;
+    }
+    dyad_kvs_response_t *resp = NULL;
+    rc = dyad_fetch(ctx, fname, &resp);
+    if (rc != 0)
+    {
+        DYAD_LOG_ERR (ctx, "dyad_fetch failed!\n");
+        if (resp != NULL)
+        {
+            dyad_free_kvs_response(resp);
+        }
+        return rc;
+    }
+    rc = dyad_pull(ctx, fname, resp);
```

Comment:
`dyad_fetch()` can finish successfully. but resp could still be NULL. `dyad_pull` should be skipped.

Reply:
I've added a conditional just before `dyad_pull` that will check for this.